src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Workspace/Mef/MefLanguageServices.cs

```diff
@@ -50,36 +57,95 @@ public bool HasServices
             get { return _services.Length > 0; }
         }
 
+        public override void Dispose()
+        {
+            ImmutableArray<IDisposable> disposableServices;
+            lock (_gate)
+            {
+                disposableServices = _ownedDisposableServices.ToImmutableArray();
+                _ownedDisposableServices.Clear();
+            }
+
+            // Take care to give all disposal parts a chance to dispose even if some parts throw exceptions.
+            List<Exception> exceptions = null;
+            foreach (var service in disposableServices)
+            {
+                MefUtilities.DisposeWithExceptionTracking(service, ref exceptions);
+            }
+
+            if (exceptions is not null)
+            {
+                throw new AggregateException(CompilerExtensionsResources.Instantiated_parts_threw_exceptions_from_IDisposable_Dispose, exceptions);
+            }
```

Comment:
Consider refactoring MefUtilities.DisposeWithExceptionTracking to just do this entire pattern of looping and throwing, since we'll need the same thing in the other place too.